Cargo.lock

```diff
@@ -2174,6 +2214,24 @@ dependencies = [
  "url",
 ]
 
+[[package]]
+name = "janus_messages"
+version = "0.5.14"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "cd1ad69e1bfa6864ad980df64d5a0e42467ae9c9d8312d6fceb13fd6203b3b95"
+dependencies = [
+ "anyhow",
+ "base64 0.21.2",
+ "derivative",
+ "hex",
+ "num_enum 0.6.1",
+ "prio",
+ "rand",
+ "serde",
+ "thiserror",
+ "url",
+]
+
```

Comment:
The presence of a second copy of `janus_messgaes` is a bad smell. I tried updating the dependency and adapting the code to pass a `HpkeConfig` to `create_hpke_config()` instead of encoding it first, and I now get an error due to mismatched crate versions.

```
error[E0308]: mismatched types
    --> integration_tests/tests/in_cluster.rs:151:17
     |
149  |             .create_hpke_config(
     |              ------------------ arguments to this method are incorrect
150  |                 account.id,
151  |                 task.collector_hpke_config().unwrap(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&HpkeConfigContents`, found `&HpkeConfig`
     |
     = note: `HpkeConfig` and `HpkeConfigContents` have similar names, but are actually distinct types
note: `HpkeConfig` is defined in crate `janus_messages`
    --> /home/david/Code/ppm/janus-ops/janus-checkout/messages/src/lib.rs:1076:1
     |
1076 | pub struct HpkeConfig {
     | ^^^^^^^^^^^^^^^^^^^^^
note: `HpkeConfigContents` is defined in crate `janus_messages`
    --> /home/david/.cargo/registry/src/index.crates.io-6f17d22bba15001f/janus_messages-0.5.14/src/lib.rs:1076:1
     |
1076 | pub struct HpkeConfig {
     | ^^^^^^^^^^^^^^^^^^^^^
     = note: perhaps two different versions of crate `janus_messages` are being used?
note: method defined here
    --> /home/david/.cargo/git/checkouts/divviup-api-9ebde8d8443dfb74/97ded06/client/src/lib.rs:322:18
     |
322  |     pub async fn create_hpke_config(
     |                  ^^^^^^^^^^^^^^^^^^
```

I think we can fix this by adding a `patch.crates-io.janus_messages.path` entry in the workspace-level `Cargo.toml`. I confirmed that `cargo publish --dry-run -p janus_messages` doesn't complain with this present.

```toml
[patch.crates-io]
janus_messages = { path = "messages" }
```

Reply:
I addressed this a different way — I'm not convinced it's a smell to have two different versions of a crate. The janus_messages in the client should be considered part of its interface, and reflects the version of janus_messages the client expects to be interacting with. Instead I just converted between the types at the api boundary. Do you still prefer patching janus_messages?

Reply:
Sure, converting sounds good to me